##**WITHOUT PIPELINE**

In [191]:
import pandas as pd
import numpy as np 

In [192]:
from sklearn.preprocessing import OneHotEncoder,StandardScaler
from sklearn.model_selection import GridSearchCV,train_test_split,RandomizedSearchCV
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline

In [193]:
df = pd.read_csv('/content/Telecom Churn Data SingTel.csv')
df.head()

State  Account Length  Area Code Phone Number International Plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  Voice mail Plan  Num of Voice mail Messages  Total Day Minutes  \
0             yes                          25              265.1   
1             yes                          26              161.6   
2              no                           0              243.4   
3              no                           0              299.4   
4              no                           0              166.7   

   Total Day Calls  Total day Charge  ...  Total Eve Calls  Total Eve Charge  \
0              110             45.07  ...               99             16.78   
1              123             27.47  ...              103             16.62   
2              114             41.38  ...              110             10.30   
3               71             50.90  ...               88              5.26   
4              113             28.34  ...              122             12.61   

   Total Night Minutes  Total Night Calls   Total Night Charge  \
0                244.7                  91               11.01   
1                254.4                 103               11.45   
2                162.6                 104                7.32   
3                196.9                  89                8.86   
4                186.9                 121                8.41   

   Total International Minutes  Total Intl  Calls  Total Intl Charge  \
0                         10.0                  3               2.70   
1                         13.7                  3               3.70   
2                         12.2                  5               3.29   
3                          6.6                  7               1.78   
4                         10.1                  3               2.73   

   Number Customer Service calls     Churn  
0                               1   False.  
1                               1   False.  
2                               0   False.  
3                               2   False.  
4                               3   False.  

[5 rows x 21 columns]

In [194]:
df.drop('Phone Number',axis=1 , inplace=True)
df.head()

State  Account Length  Area Code International Plan Voice mail Plan  \
0    KS             128        415                 no             yes   
1    OH             107        415                 no             yes   
2    NJ             137        415                 no              no   
3    OH              84        408                yes              no   
4    OK              75        415                yes              no   

   Num of Voice mail Messages  Total Day Minutes  Total Day Calls  \
0                          25              265.1              110   
1                          26              161.6              123   
2                           0              243.4              114   
3                           0              299.4               71   
4                           0              166.7              113   

   Total day Charge  Total Eve Minutes  Total Eve Calls  Total Eve Charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total Night Minutes  Total Night Calls   Total Night Charge  \
0                244.7                  91               11.01   
1                254.4                 103               11.45   
2                162.6                 104                7.32   
3                196.9                  89                8.86   
4                186.9                 121                8.41   

   Total International Minutes  Total Intl  Calls  Total Intl Charge  \
0                         10.0                  3               2.70   
1                         13.7                  3               3.70   
2                         12.2                  5               3.29   
3                          6.6                  7               1.78   
4                         10.1                  3               2.73   

   Number Customer Service calls     Churn  
0                               1   False.  
1                               1   False.  
2                               0   False.  
3                               2   False.  
4                               3   False.

In [195]:
df.isna().sum()

State                             0
Account Length                    0
Area Code                         0
International Plan                0
Voice mail Plan                   0
Num of Voice mail Messages        0
Total Day Minutes                 0
Total Day Calls                   0
Total day Charge                  0
Total Eve Minutes                 0
Total Eve Calls                   0
Total Eve Charge                  0
Total Night Minutes               0
Total Night Calls                 0
Total Night Charge                0
Total International Minutes       0
Total Intl  Calls                 0
Total Intl Charge                 0
Number Customer Service calls     0
Churn                             0
dtype: int64

In [196]:
#Transforming the Categorical Columns into Numerical columns
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder()
df['Churn']=label_encoder.fit_transform(df['Churn'])
df['State'] = label_encoder.fit_transform(df['State'])
df['International Plan'] = label_encoder.fit_transform(df['International Plan'])
df['Voice mail Plan'] = label_encoder.fit_transform(df['Voice mail Plan'])
df.head()

State  Account Length  Area Code  International Plan  Voice mail Plan  \
0     16             128        415                   0                1   
1     35             107        415                   0                1   
2     31             137        415                   0                0   
3     35              84        408                   1                0   
4     36              75        415                   1                0   

   Num of Voice mail Messages  Total Day Minutes  Total Day Calls  \
0                          25              265.1              110   
1                          26              161.6              123   
2                           0              243.4              114   
3                           0              299.4               71   
4                           0              166.7              113   

   Total day Charge  Total Eve Minutes  Total Eve Calls  Total Eve Charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total Night Minutes  Total Night Calls   Total Night Charge  \
0                244.7                  91               11.01   
1                254.4                 103               11.45   
2                162.6                 104                7.32   
3                196.9                  89                8.86   
4                186.9                 121                8.41   

   Total International Minutes  Total Intl  Calls  Total Intl Charge  \
0                         10.0                  3               2.70   
1                         13.7                  3               3.70   
2                         12.2                  5               3.29   
3                          6.6                  7               1.78   
4                         10.1                  3               2.73   

   Number Customer Service calls   Churn  
0                               1      0  
1                               1      0  
2                               0      0  
3                               2      0  
4                               3      0

In [197]:
x = df.drop('Churn' , 1)
y = df['Churn']
y.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  """Entry point for launching an IPython kernel.


0    0
1    0
2    0
3    0
4    0
Name: Churn, dtype: int64

In [198]:
y.value_counts(normalize=True)

0    0.8586
1    0.1414
Name: Churn, dtype: float64

In [199]:
x_train , x_test , y_train , y_test = train_test_split(x, y , test_size=0.2 , random_state=0 , stratify=y)

In [200]:
df.info

<bound method DataFrame.info of       State  Account Length  Area Code  International Plan  Voice mail Plan  \
0        16             128        415                   0                1   
1        35             107        415                   0                1   
2        31             137        415                   0                0   
3        35              84        408                   1                0   
4        36              75        415                   1                0   
...     ...             ...        ...                 ...              ...   
4995     11              50        408                   0                1   
4996     49             152        415                   0                0   
4997      7              61        415                   0                0   
4998      7             109        510                   0                0   
4999     46              86        415                   0                1   

      Num of Voice 

In [201]:
cat_cols = ['State' , 'International Plan' , 'Voice mail Plan'] #categorical columns
#numerical columns
num_cols = ['Account Length','Area Code', 'Num of Voice mail Messages','Total Day Minutes', 'Total Day Calls', 'Total day Charge','Total Eve Minutes', 'Total Eve Calls', 'Total Eve Charge', 'Total Night Minutes', 'Total Night Calls ', 'Total Night Charge', 'Total International Minutes', 'Total Intl  Calls', 'Total Intl Charge','Number Customer Service calls ']

In [202]:
df.columns

Index(['State', 'Account Length', 'Area Code', 'International Plan',
       'Voice mail Plan', 'Num of Voice mail Messages', 'Total Day Minutes',
       'Total Day Calls', 'Total day Charge', 'Total Eve Minutes',
       'Total Eve Calls', 'Total Eve Charge', 'Total Night Minutes',
       'Total Night Calls ', 'Total Night Charge',
       'Total International Minutes', 'Total Intl  Calls', 'Total Intl Charge',
       'Number Customer Service calls ', 'Churn'],
      dtype='object')

Rescaling the Numerical Columns to Standarization >>> N(0,1) 

In [203]:
ss = StandardScaler()
x_train_num = ss.fit_transform(x_train[num_cols])
x_train_num[:5]

array([[-2.27947876, -0.51953119,  2.28117   ,  0.63595829,  0.40361454,
         0.63622895, -0.63626781,  0.75353204, -0.6367876 , -0.20591547,
         0.87419056, -0.20440472, -0.05916254, -0.57579076, -0.05584197,
        -0.43173921],
       [-0.4772175 , -0.51953119,  1.4748568 , -0.22638334,  2.3647392 ,
        -0.22677185, -0.20801221,  0.25171262, -0.20794481, -0.14064155,
         0.17316567, -0.13847197, -0.38497738,  0.62942276, -0.39106814,
         1.86780386],
       [ 1.85070328, -0.68531319, -0.57757681,  0.22531942, -0.14952318,
         0.22558989, -0.6223246 , -0.30028874, -0.62272718, -0.66283288,
        -2.28042147, -0.66153848,  0.41145889, -0.17405292,  0.41347466,
        -0.43173921],
       [ 0.42391312,  1.73036741, -0.57757681, -1.29964403,  1.10760801,
        -1.29948404,  0.83175255,  0.15134874,  0.83252621, -0.58766898,
        -1.22888413, -0.58681469,  1.3165001 ,  0.22768492,  1.31188078,
        -0.43173921],
       [-2.40463579, -0.51953119, -0

In [204]:
x_test_num = ss.fit_transform(x_test[num_cols])
x_test_num[:5]

array([[ 0.0767865 , -0.51770713, -0.55249068,  1.88637203, -1.28864147,
         1.88622542, -1.30796561,  0.51488099, -1.30757512, -0.39141785,
        -0.12249989, -0.39198447,  0.08881271, -1.05239653,  0.09487578,
         0.30556437],
       [-0.93288088, -0.51770713, -0.55249068, -1.17081   , -0.67629056,
        -1.17046644, -0.38019226, -0.41366903, -0.37945403, -0.37950671,
         2.44954389, -0.3787498 ,  1.39488196, -1.05239653,  1.39840982,
        -0.45644153],
       [ 0.12856431, -0.51770713,  2.26400907,  0.76510289,  0.19120657,
         0.76558071,  0.70381735, -0.10415236,  0.70297571, -1.21924191,
        -0.57638997, -1.22135714,  0.12509241,  0.23994469,  0.12175277,
         0.30556437],
       [ 0.33567557, -0.68386715, -0.55249068,  0.41856178, -0.47217358,
         0.41871449,  0.79061527,  1.8045338 ,  0.79147625,  0.66470302,
        -1.88762797,  0.66678916,  0.45160972, -0.62161613,  0.45771515,
         0.30556437],
       [-0.18210257,  1.73732165, -0

We'll concatenate/Add the X_train_num with X_train of the categorical columns to finally give the Final x_train_out output dataframe

In [205]:
x_train_out = pd.DataFrame(np.concatenate([x_train_num , x_train[cat_cols]],axis=1))
x_train_out.head()

0         1         2         3         4         5         6   \
0 -2.279479 -0.519531  2.281170  0.635958  0.403615  0.636229 -0.636268   
1 -0.477218 -0.519531  1.474857 -0.226383  2.364739 -0.226772 -0.208012   
2  1.850703 -0.685313 -0.577577  0.225319 -0.149523  0.225590 -0.622325   
3  0.423913  1.730367 -0.577577 -1.299644  1.107608 -1.299484  0.831753   
4 -2.404636 -0.519531 -0.577577 -0.192786 -0.199808 -0.192735 -1.381233   

         7         8         9         10        11        12        13  \
0  0.753532 -0.636788 -0.205915  0.874191 -0.204405 -0.059163 -0.575791   
1  0.251713 -0.207945 -0.140642  0.173166 -0.138472 -0.384977  0.629423   
2 -0.300289 -0.622727 -0.662833 -2.280421 -0.661538  0.411459 -0.174053   
3  0.151349  0.832526 -0.587669 -1.228884 -0.586815  1.316500  0.227685   
4 -0.300289 -1.381990  0.889895 -0.678079  0.890079  1.425105 -0.174053   

         14        15    16   17   18  
0 -0.055842 -0.431739  18.0  0.0  1.0  
1 -0.391068  1.867804  33.0  0.0  1.0  
2  0.413475 -0.431739  13.0  0.0  0.0  
3  1.311881 -0.431739  10.0  0.0  0.0  
4  1.419153 -1.198254  35.0  1.0  0.0

In [206]:
x_test_out = pd.DataFrame(np.concatenate([x_test_num,x_test[num_cols]],axis=1))
x_test_out.head()

0         1         2         3         4         5         6   \
0  0.076786 -0.517707 -0.552491  1.886372 -1.288641  1.886225 -1.307966   
1 -0.932881 -0.517707 -0.552491 -1.170810 -0.676291 -1.170466 -0.380192   
2  0.128564 -0.517707  2.264009  0.765103  0.191207  0.765581  0.703817   
3  0.335676 -0.683867 -0.552491  0.418562 -0.472174  0.418714  0.790615   
4 -0.182103  1.737322 -0.552491 -0.076756  0.599441 -0.076504 -0.088937   

         7         8         9   ...     22     23     24     25     26  \
0  0.514881 -1.307575 -0.391418  ...  130.8  111.0  11.12  181.8   99.0   
1 -0.413669 -0.379454 -0.379507  ...  178.9   93.0  15.21  182.4  150.0   
2 -0.104152  0.702976 -1.219242  ...  235.1   99.0  19.98  140.1   90.0   
3  1.804534  0.791476  0.664703  ...  239.6  136.0  20.37  235.0   64.0   
4  0.566467 -0.088991 -1.088219  ...  194.0  112.0  16.49  146.7  108.0   

      27    28   29    30   31  
0   8.18  10.5  2.0  2.84  2.0  
1   8.21  14.1  2.0  3.81  1.0  
2   6.30  10.6  5.0  2.86  2.0  
3  10.58  11.5  3.0  3.11  2.0  
4   6.60   5.9  4.0  1.59  1.0  

[5 rows x 32 columns]

We've Now the test and the training data Ready !! \
Moving to the **Random Forest Classifier** by initiating the **rfc** and put the paramaters multiple choice to find the optimal one

In [207]:
rfc = RandomForestClassifier(random_state=0)
params = {
    'n_estimators' : [10,20],
    'max_depth' : [2,3]
}

Moving to the **GridSearchCV** which we're going to use it 

In [208]:
gs = GridSearchCV(estimator=rfc , param_grid=params , cv=5 , n_jobs=-1)

**n_jobs** : mean how many compututions we want , is the Number of jobs to run in parallel , **-1** means we using all processus so all the **combinations** between parameters .

In [209]:
gs.fit(x_train_out , y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=0), n_jobs=-1,
             param_grid={'max_depth': [2, 3], 'n_estimators': [10, 20]})

let's Look at the **Result** of GridSearchCV

In [210]:
pd.DataFrame(gs.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.057044      0.007084         0.008775        0.002433   
1       0.092511      0.007418         0.008691        0.000534   
2       0.056760      0.001963         0.006863        0.000390   
3       0.100234      0.008966         0.008156        0.001423   

  param_max_depth param_n_estimators                                params  \
0               2                 10  {'max_depth': 2, 'n_estimators': 10}   
1               2                 20  {'max_depth': 2, 'n_estimators': 20}   
2               3                 10  {'max_depth': 3, 'n_estimators': 10}   
3               3                 20  {'max_depth': 3, 'n_estimators': 20}   

   split0_test_score  split1_test_score  split2_test_score  split3_test_score  \
0            0.85875            0.86000            0.85875            0.85875   
1            0.85875            0.86250            0.86375            0.86125   
2            0.89000            0.88625            0.88625            0.88375   
3            0.87875            0.89000            0.88375            0.88375   

   split4_test_score  mean_test_score  std_test_score  rank_test_score  
0            0.85750          0.85875        0.000791                4  
1            0.86375          0.86200        0.001871                3  
2            0.88750          0.88675        0.002031                1  
3            0.88500          0.88425        0.003588                2

we gonna see the several parameters which the model been evaluating , and in the end it'll giving us the **rank_tes_score** according to the model \
Or we can do simply **gs.best_params_** it will give us the optimal parameters for the model

In [211]:
gs.best_params_

{'max_depth': 3, 'n_estimators': 10}

we gonna utilize the previous optimal parameters that we obtain and fitting over the x and y training .

In [212]:
rf = RandomForestClassifier(random_state=0 , **gs.best_params_)

In [213]:
rf.fit(x_test_out , y_test)

RandomForestClassifier(max_depth=3, n_estimators=10, random_state=0)

In [214]:
rf.score(x_test_out , y_test)

0.912

the score the we obtain is **91.2%** occurancy 

##**WITH PIPELINE**

In [215]:
x_train.columns

Index(['State', 'Account Length', 'Area Code', 'International Plan',
       'Voice mail Plan', 'Num of Voice mail Messages', 'Total Day Minutes',
       'Total Day Calls', 'Total day Charge', 'Total Eve Minutes',
       'Total Eve Calls', 'Total Eve Charge', 'Total Night Minutes',
       'Total Night Calls ', 'Total Night Charge',
       'Total International Minutes', 'Total Intl  Calls', 'Total Intl Charge',
       'Number Customer Service calls '],
      dtype='object')

In [216]:
ct = ColumnTransformer([('ss' , StandardScaler() , num_cols)] , remainder='passthrough')

we'll fit our X train

In [217]:
ct.fit(x_train)

ColumnTransformer(remainder='passthrough',
                  transformers=[('ss', StandardScaler(),
                                 ['Account Length', 'Area Code',
                                  'Num of Voice mail Messages',
                                  'Total Day Minutes', 'Total Day Calls',
                                  'Total day Charge', 'Total Eve Minutes',
                                  'Total Eve Calls', 'Total Eve Charge',
                                  'Total Night Minutes', 'Total Night Calls ',
                                  'Total Night Charge',
                                  'Total International Minutes',
                                  'Total Intl  Calls', 'Total Intl Charge',
                                  'Number Customer Service calls '])])

In [218]:
ct.transformers_

[('ss',
  StandardScaler(),
  ['Account Length',
   'Area Code',
   'Num of Voice mail Messages',
   'Total Day Minutes',
   'Total Day Calls',
   'Total day Charge',
   'Total Eve Minutes',
   'Total Eve Calls',
   'Total Eve Charge',
   'Total Night Minutes',
   'Total Night Calls ',
   'Total Night Charge',
   'Total International Minutes',
   'Total Intl  Calls',
   'Total Intl Charge',
   'Number Customer Service calls ']),
 ('remainder', 'passthrough', [0, 3, 4])]

it passes through all the categorical cols which we've mentioned , \
we'll Creating the Pipeline with our training data and then applying RandomForestClassifier

In [219]:
from scipy.sparse.construct import rand
pipe = Pipeline([
                 ('ct' , ct),
                 ('model' , RandomForestClassifier(random_state=0 , n_jobs=-1))
])

In [220]:
params_new = {
    'model__n_estimators' : [10,20],
    'model__max_depth': [2,3],
    'ct__ss__with_mean' : [True,False]
}

In [221]:
gs_new = GridSearchCV(estimator=pipe , param_grid=params_new , n_jobs=-1)

In [222]:
x_train.head()

State  Account Length  Area Code  International Plan  Voice mail Plan  \
1140     18               9        415                   0                1   
944      33              81        415                   0                1   
60       13             174        408                   0                0   
3696     10             117        510                   0                0   
4162     35               4        415                   1                0   

      Num of Voice mail Messages  Total Day Minutes  Total Day Calls  \
1140                          39              214.1              108   
944                           28              167.9              147   
60                             0              192.1               97   
3696                           0              110.4              122   
4162                           0              169.7               96   

      Total day Charge  Total Eve Minutes  Total Eve Calls  Total Eve Charge  \
1140             36.40              169.2              115             14.38   
944              28.54              190.7              105             16.21   
60               32.66              169.9               94             14.44   
3696             18.77              242.9              103             20.65   
4162             28.85              131.8               94             11.20   

      Total Night Minutes  Total Night Calls   Total Night Charge  \
1140                189.7                 117                8.54   
944                 193.0                 103                8.69   
60                  166.6                  54                7.50   
3696                170.4                  75                7.67   
4162                245.1                  86               11.03   

      Total International Minutes  Total Intl  Calls  Total Intl Charge  \
1140                         10.1                  3               2.73   
944                           9.2                  6               2.48   
60                           11.4                  4               3.08   
3696                         13.9                  5               3.75   
4162                         14.2                  4               3.83   

      Number Customer Service calls   
1140                               1  
944                                4  
60                                 1  
3696                               1  
4162                               0

then fitting our data on new gs

In [223]:
x_train.head()

State  Account Length  Area Code  International Plan  Voice mail Plan  \
1140     18               9        415                   0                1   
944      33              81        415                   0                1   
60       13             174        408                   0                0   
3696     10             117        510                   0                0   
4162     35               4        415                   1                0   

      Num of Voice mail Messages  Total Day Minutes  Total Day Calls  \
1140                          39              214.1              108   
944                           28              167.9              147   
60                             0              192.1               97   
3696                           0              110.4              122   
4162                           0              169.7               96   

      Total day Charge  Total Eve Minutes  Total Eve Calls  Total Eve Charge  \
1140             36.40              169.2              115             14.38   
944              28.54              190.7              105             16.21   
60               32.66              169.9               94             14.44   
3696             18.77              242.9              103             20.65   
4162             28.85              131.8               94             11.20   

      Total Night Minutes  Total Night Calls   Total Night Charge  \
1140                189.7                 117                8.54   
944                 193.0                 103                8.69   
60                  166.6                  54                7.50   
3696                170.4                  75                7.67   
4162                245.1                  86               11.03   

      Total International Minutes  Total Intl  Calls  Total Intl Charge  \
1140                         10.1                  3               2.73   
944                           9.2                  6               2.48   
60                           11.4                  4               3.08   
3696                         13.9                  5               3.75   
4162                         14.2                  4               3.83   

      Number Customer Service calls   
1140                               1  
944                                4  
60                                 1  
3696                               1  
4162                               0

In [224]:
gs_new.fit(x_train,y_train)

GridSearchCV(estimator=Pipeline(steps=[('ct',
                                        ColumnTransformer(remainder='passthrough',
                                                          transformers=[('ss',
                                                                         StandardScaler(),
                                                                         ['Account '
                                                                          'Length',
                                                                          'Area '
                                                                          'Code',
                                                                          'Num '
                                                                          'of '
                                                                          'Voice '
                                                                          'mail '
                                          

In [225]:
gs_new.best_params_

{'ct__ss__with_mean': True, 'model__max_depth': 3, 'model__n_estimators': 10}

In [226]:
gs_new.best_score_

0.8867499999999999